### ISSUE TITLE
LLVM does not build with libc++ built without threading support.

### ISSUE THREAD

**Issue by veluca93** [llvm/llvm-project#76588]
As per title, if one builds `libc++` without threading support, the resulting `libc++` cannot then be used to build `llvm` (or other llvm projects).

This is also true if one passes `-DLLVM_ENABLE_THREADS=OFF` to the CMake build script.

The error messages are numerous, generally coming from unguarded usage of C++ thread primitives all over the codebase.

**Comment by mordante** [comment]
In general the LLVM project requires C++17 support, which means threading support. I'm not sure how well the `DLLVM_ENABLE_THREADS` option is supported in subprojects. At least this is not a libc++ issue.

**Comment by veluca93** [comment]
I agree this is not a libc++ issue, but I don't think I can change the tag
myself...

Although there might be a small libc++ issue in that for reasons I do not
understand, including <future> in a libc++ built without thread support did
not error out in the expected way, but rather caused errors inside thread.h.

**Comment by mordante** [comment]
> I agree this is not a libc++ issue, but I don't think I can change the tag myself... Although there might be a small libc++ issue in that for reasons I do not understand, including <future> in a libc++ built without thread support did not error out in the expected way, but rather caused errors inside thread.h.

That's odd, there is a a guard
https://github.com/llvm/llvm-project/blob/main/libcxx/include/future#L400
Can you post a minimal example to reproduce the issue?

**Comment by veluca93** [comment]
I also saw the guard and was hence fairly confused :-)

I get the following behaviour with different `include`s (just a single `.cc` file with `#include <whatever>`, compiled with `./build/llvm-host/bin/clang++ -c -o /dev/null weird.cc` for a clang that points to a WASM sysroot with a libc++ with threads disabled):

- `mutex`: compilation succeeds
- `shared_mutex`: 
```
/home/luca/wasm-compilers/build/sysroot/include/c++/v1/shared_mutex:151:6: error: "<shared_mutex> is not supported since libc++ has been configured without support for threads."
```
- `thread` or `future`:
```
In file included from weird.cc:1:
In file included from /home/luca/wasm-compilers/build/sysroot/include/c++/v1/thread:93:
In file included from /home/luca/wasm-compilers/build/sysroot/include/c++/v1/__thread/jthread.h:18:
/home/luca/wasm-compilers/build/sysroot/include/c++/v1/__thread/thread.h:55:34: error: unknown type name 'condition_variable'
   55 |   void notify_all_at_thread_exit(condition_variable*, mutex*);
      |                                  ^
```

I'm not quite sure what the difference would be between these headers, at first inspection the guard seems the same in all of them.

**Comment by mordante** [comment]
Thanks for the additional information. This is a libc++ issue, but not really related to your original issue. I've created #76620.

**Comment by veluca93** [comment]
FWIW, if there's desire to support building without threads, I managed to get a fair chunk of llvm to compile by mocking out the relevant C++17 thread APIs, similarly to how `llvm::thread` is defined in `llvm/Support/thread.h`, and then replacing references to the std versions to references to the `llvm` ones.

That said, I am not sure there's much value in doing this if i.e. there's no builders that build with LLVM_ENABLE_THREADS disabled...
